### DISCUSSION TITLE
Issues implementing CastShape function for custom shape

### DISCUSSION THREAD

**Discussion by JuanDiegoMontoya** [jrouwe/JoltPhysics#1381] [Q&A] [answered]
## Background

I'm using Jolt to add rigid body dynamics to my blocky voxel game, and I'm implementing a custom shape to represent the world. Thanks to Jolt's excellent docs, this hasn't been too difficult.

For objects with the discrete motion quality, this was surprisingly easy. I only had to register a CollideShape function. The strategy I used here was to loop over all the blocks spanned by the AABox of the other shape, make a translated box, then call `CollideDispatch::sCollideShapeVsShape` with the box and the other shape. This worked great.

When I decided to try `CharacterVirtual`, I found that I needed to implement CastShape. No problem, I'd just use the same strategy as before, except with the AABox of the cast shape (the union of the shape's world bounds and its bounds plus the direction of the cast). Then I would just forward it to `CollisionDispatch::sCastShapeVsShapeLocalSpace` to resolve box-vs-swept-other-shape collisions.

## The issue(s)

Unfortunately, I encountered two issues that I think stem from my implementation of CastShape being broken. Starting with the simpler:

1. Objects with the LinearCast motion quality still collide with my shape fine, but tunnel through it at high speeds.
2. My `CharacterVirtual` has trouble staying in contact with my custom shape specifically. The character will tunnel into it a little bit before getting pushed out the next frame. This results in goofy "bouncing" when standing still or running into a wall. A video of the effect follows:

https://github.com/user-attachments/assets/ddfce48c-99b7-428e-b8a3-2b3492098809

## What I've tried + observations

- My `CharacterVirtual` interacts well with shapes in the scene that aren't my custom shape, e.g. not being able to tunnel through them at high speeds.
- Setting `mCharacterPadding` to 0 in `CharacterVirtual` fixes with the bounciness issue. It doesn't help with tunneling though, plus it seems like a hack.
- Checking that the region of voxels being checked makes sense in a given situation. For example, checking that moving extremely quickly makes a large AABox for the cast shape.
- The value of `inCenterOfMassTransform` that I pass to `sCastShapeVsShapeLocalSpace` doesn't seem to have any effect. Is this because the body of the voxel world has the static movement type?

I'm fairly certain the issues with my character are a byproduct of my CastShape function being broken, but I don't know how. Either way, regular rigid bodies tunneling through my shape seems like a fundamental issue that needs to be fixed before moving on.

Are there any common gotchas when implementing a CastShape function? Maybe there's someone else who tried making a custom shape for voxels with Jolt? Surely I'm not the only one 😅. If not, this issue could serve as a resource for others in the future (I did search the issues and found none that seemed like my particular problem).

## Code

- [CastShape function for my voxel world](https://github.com/JuanDiegoMontoya/Frogfood/blob/45275d1874af6e84e74e446bcdc56c41594eb060/src/Voxels/Physics/TwoLevelGridShape.cpp#L66-L129)
- [CharacterVirtual creation](https://github.com/JuanDiegoMontoya/Frogfood/blob/45275d1874af6e84e74e446bcdc56c41594eb060/src/Voxels/Physics/Physics.cpp#L201-L209)

**Comment by jrouwe** [comment, marked as answer]
I would suggest to create a [debug probe](https://github.com/jrouwe/JoltPhysics/blob/e0e434f59c2ff6816e07af7f6f6fcf9726e2aa0f/Samples/SamplesApp.cpp#L1364-L1474) like the Samples to test your `CastShape` function.

Looking at the code, I think the problem is that you're not making the cast shape relative to the box: https://github.com/JuanDiegoMontoya/Frogfood/blob/45275d1874af6e84e74e446bcdc56c41594eb060/src/Voxels/Physics/TwoLevelGridShape.cpp#L117

This `inCastShape` should be replaced by something like:

```
ShapeCast shape_cast = inShapeCast.PostTranslated({-x - 0.5f, -y - 0.5f, -z - 0.5f});
```

`sCastShapeVsShapeLocalSpace` takes a local space cast as the function name suggests, the center of mass transform that is being passed in is only used to transform the normals and contact points to world space.

Another suggestion that I have is to not new the `BoxShape` but do:

```
BoxShape boxShape(...);
boxShape.SetEmbedded();
```

That saves 1 dynamic memory allocation per cast.

Finally, you can improve the 'broad phase check' by not testing the bounding box of the shape + shape moved by cast direction, but by adding the size of the bounding box of the cast shape to your target box and casting a ray against the box. Example here: https://github.com/jrouwe/JoltPhysics/blob/e0e434f59c2ff6816e07af7f6f6fcf9726e2aa0f/Jolt/Physics/Collision/Shape/CompoundShapeVisitors.h#L178-L190

**Reply by JuanDiegoMontoya** [reply]
Hi, apologies for the delay. I was implementing your suggestions.

- Translating the shape cast by the opposite of the voxel position fixed CCD. No more tunneling!
- I implemented a debug renderer and created a probe as you recommended. It seems to collide correctly and the collision normals are correct. A video of it follows (the probe is placed in front of the player and moves left-to-right in the player's local space):

https://github.com/user-attachments/assets/88110ab9-6bc4-4b2e-adae-57859dc81e6b

I also found a (presumably) floating-point error in my "broadphase" code that sometimes caused some boxes to not be checked, leading to the player becoming stuck. A simple epsilon fixed it.

The last wrinkle I have is that capsule and sphere player shapes have minor ghost collision issues when moving over the seam between two voxels. The effect isn't large (the player's Y-position varies by around 2e-2 when walking on flat ground) but it is noticeable upon close inspection. Without `mEnhancedInternalEdgeRemoval` set to true *or* reducing `mPredictiveContactDistance`, the issue is considerably worse.

That being said, if I give the player a box shape, movement is smooth as butter. I also suspect that the behavior with capsules and spheres is considered normal for the geometry they're encountering.

Again, thanks for taking the time to help.

**Comment by JuanDiegoMontoya** [comment]
Thanks for the suggestions. I'll implement these then let you know how it went.
